File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        if(nums.size() == 0)
            return 0;
        
        vector<int> LIS;
        
        for(auto& num : nums)
        {
            auto it=lower_bound(LIS.begin(),LIS.end(),num);
            
            if(it == LIS.end())
                LIS.push_back(num);
            else
                *it=num;
        }
        
        return LIS.size();
    }
};

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
    private int getLowerBound(List<Integer> increasingSubsequence, int target){
        int left = 0, right = increasingSubsequence.size()-1;
        while(left != right){
            int mid = (left + right)/2;
            if(increasingSubsequence.get(mid) < target){
                left = mid + 1;
            } else{
                right = mid;
            }
        }
        return left;
    }
    
    public int lengthOfLIS(int[] nums) {
        List<Integer> increasingSubsequence = new ArrayList<>();
        for(int num : nums){
            if(increasingSubsequence.size() == 0 || 
               increasingSubsequence.get(increasingSubsequence.size() - 1) < num){
                increasingSubsequence.add(num);
            } else {
                int lower_bound_index = getLowerBound(increasingSubsequence, num);
                increasingSubsequence.set(lower_bound_index, num); 
            }
        }
        return increasingSubsequence.size();
    }
}